File: 4.py
#!/usr/bin/env python

"""This script solves the Project Euler problem "Largest palindrome product".
The problem is: Find the largest palindrome made from the product of two
3-digit numbers.
"""

import argparse


def main(args):
    """Largest palindrome product"""

    high_num = pow(10, args.digits)

    max = 0

    for num1 in reversed(range(high_num)):
        for num2 in reversed(range(high_num)):
            if num2 <= num1:
                product = num1 * num2
                if is_palindrome(product) and product > max:
                    max = product

    print(max)


def is_palindrome(number):
    """Check if number is palindrome"""

    return number == int(str(number)[::-1])

if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Largest palindrome product')
    parser.add_argument(
        'digits', metavar='DIGITS', type=int, default=3, nargs='?',
        help='Number of digits in numbers to multiply')
    args = parser.parse_args()

    main(args)

File: 4.js
#!/usr/bin/env node

// Largest palindrome product

var program = require('commander');

program
    .version('0.1.0')
    .description('Largest palindrome product')
    .option('-d, --digits <int>',
        'The number of digits in the numbers to be multiplied', Number, 3)
    .parse(process.argv);

var digits = program.digits;
var high_num = '9'.repeat(digits);
var max = 0;
for (var num1 = high_num; num1 > 0; num1--) {
    for (var num2 = high_num; num2 > 0; num2--) {
        if (num2 > num1) {
            continue;
        }
        var product = num1 * num2;
        if (product > max && is_palindrome(product)) {
            max = product;
        }
    }
}

console.log(max);

function is_palindrome(number) {
    return number.toString().split('').reverse('').join('') == number;
}
